packages/odyssey-react-mui/src/labs/FileUpload.tsx

```diff
@@ -0,0 +1,294 @@
+import {
+  useOdysseyDesignTokens,
+  DesignTokens,
+} from "../OdysseyDesignTokensContext";
+import { Support } from "../Typography";
+
+export const fileUploadTypes = ["single", "multiple"] as const;
+export const fileUploadVariants = [
+  "button",
+  "dragAndDrop",
+  "dragAndDropWithIcon",
+] as const;
+
+const BaseInputWrapper = styled.div({
+  position: "relative",
+  alignSelf: "flex-start",
+
+  input: {
+    position: "absolute",
+    width: "100%",
+    height: "100%",
+    opacity: 0,
+  },
+});
+
+const InputContainer = styled(BaseInputWrapper)<{
+  odysseyDesignTokens: DesignTokens;
+}>(({ odysseyDesignTokens }) => ({
+  display: "flex",
+  alignSelf: "unset",
+  alignItems: "center",
+  justifyContent: "center",
+  padding: `${odysseyDesignTokens.Spacing6} ${odysseyDesignTokens.Spacing3}`,
+  border: `1px dashed ${odysseyDesignTokens.HueNeutral300}`,
+  borderRadius: odysseyDesignTokens.BorderRadiusMain,
+  transition: `border-color ${odysseyDesignTokens.TransitionTimingMain}, box-shadow ${odysseyDesignTokens.TransitionTimingMain}`,
+
+  "&:hover": {
+    borderColor: odysseyDesignTokens.HueNeutral700,
+  },
+
+  "&:has(input:focus)": {
+    borderStyle: "solid",
+    borderColor: odysseyDesignTokens.FocusOutlineColorPrimary,
+    boxShadow: `0 0 0 1px ${odysseyDesignTokens.FocusOutlineColorPrimary}`,
+    outline: `${odysseyDesignTokens.FocusOutlineWidthMain} ${odysseyDesignTokens.FocusOutlineStyle} transparent`,
+    outlineOffset: odysseyDesignTokens.FocusOutlineOffsetTight,
+  },
+
+  "&:has(input:disabled)": {
+    backgroundColor: odysseyDesignTokens.HueNeutral50,
+    border: `1px solid ${odysseyDesignTokens.BorderColorDisabled}`,
+    color: odysseyDesignTokens.TypographyColorDisabled,
+
+    "&:hover": {
+      borderColor: odysseyDesignTokens.BorderColorDisabled,
+    },
+  },
+}));
+
+const ButtonAndInfoContainer = styled.div({
+  display: "flex",
+  flexDirection: "column",
+  alignItems: "center",
+  justifyContent: "center",
+});
+
+export type FileUploadProps = {
+  /**
+   * an array of file types the user is able to upload. @see https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/accept#unique_file_type_specifiers for examples
+   */
+  acceptedFileTypes?: string[];
+  /**
+   * The label for the `input` element.
+   */
+  label: string;
+  /**
+   * Function that is called when the list of ifles to upload is changed
+   */
+  onChange: (files: File[]) => void;
+  /**
+   * Either `single` or `multiple`. If `multiple`, multiple files can be uploaded
+   */
+  type?: (typeof fileUploadTypes)[number];
+  /**
+   * Either `button` or ``. If `multiple`, multiple files can be uploaded
+   */
+  variant: (typeof fileUploadVariants)[number];
+} & Pick<
+  FieldComponentProps,
+  | "errorMessage"
+  | "hint"
+  | "HintLinkComponent"
+  | "id"
+  | "isDisabled"
+  | "isOptional"
+>;
+
+type RemoveFileFromFilesToUploadList = (name: string) => void;
```

Comment:
I don't understand the name of this type.

We can just put this inline down below:
```ts
  const removeFileFromFilesToUploadList =
    useCallback<(name: string) => void>(
      (name) => {
        const deletedFileFilteredOut = filesToUpload.filter(
          (file) => file.name !== name,
        );
        setFilesToUpload(deletedFileFilteredOut);
      },
      [filesToUpload],
    );
```